src/python/pants/engine/platform.py

```diff
@@ -0,0 +1,31 @@
+# Copyright 2019 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from pants.engine.rules import rule
+from pants.util.memo import memoized_classproperty, memoized_property
+from pants.util.objects import enum
+from pants.util.osutil import all_normalized_os_names, get_normalized_os_name
+
+
+class Platform(enum(all_normalized_os_names())):
+
+  # TODO: try to turn all of these accesses into v2 dependency injections!
+  @memoized_classproperty
+  def current(cls):
+    return cls(get_normalized_os_name())
+
+  @memoized_property
+  def runtime_lib_path_env_var(self):
+    return self.resolve_for_enum_variant({
+      'darwin': 'DYLD_LIBRARY_PATH',
+      'linux': 'LD_LIBRARY_PATH',
+    })
+
+
+@rule(Platform, [])
```

Comment:
Would you mind adding a note here that although this is currently a singleton, we plan to convert it into a `RootRule` at some point so that it is specified by a caller?

Alternatively, it might be easier in a fast-follow PR? You'd have to fix up a few callsites to actually pass it in as a Param [here](https://github.com/pantsbuild/pants/blob/a24307e574947327bc7cad2321e8ad5528b972b9/src/python/pants/init/engine_initializer.py#L170-L195) (and probably wherever you moved the platform-producing code from).

Reply:
I think I will just add a note here for now. It might make sense to keep this singleton and add Params for user configured --target-platform and --execution-platforms. I'm not sure if it makes sense to roll all of those into the one Platform type or have separate types. I think rules might still want to know what the underlying arch is even if the user has configured the execution or target platforms to be something else.

Reply:
> I think rules might still want to know what the underlying arch is even if the user has configured the execution or target platforms to be something else.

I don't think that is compatible with the other constraints/goals mentioned in the design doc. IMO, it is absolutely a goal to support multi-platform testing within a single run, so this should _not_ stay a singleton indefinitiely. Having options is not incompatible with the param arriving via `RootRule`: that would just be the default at the root of the graph.

Reply:
What I mean is that in addition to rules having the requested execution platform specified, for testing in your example, it still seems useful to provide a type that reflects the actual local platform. So I guess that would manifest with a RootRule(ExecutionPlatform) or something like that in addition to this singleton. I will add a TODO for now and we can hash it out in another ticket.

Reply:
> for testing in your example, it still seems useful to provide a type that reflects the actual local platform.

It's not clear why. If the param that comes in is Linux, the rules should prepare for Linux, regardless of which platform we are running on.

Having the param specified as a `RootRule` would result in using the _default_ specified by the option (probably the current platform). But you don't want to _hardcode_ it to the local platform (in the long term), which is what a singleton does.

Reply:
I see your point. I don't have a concrete use case in mind. I guess if one comes up we can deal with it then.

Reply:
I see your point. I don't have a concrete use case in mind. If it comes up in the future we can deal with it then.